### PR TITLE
multicore: remove coherence units

### PR DESCRIPTION
**Pull request by devins2518**
The coherence units add unnecessary latency and logic to bus requests. This PR removes the middle man and directly integrates the `bus_controller` into `l1_cache`. With this change, it will be easier to debug coherence/deadlock issues, and (this is purely speculative) reduce the ability to deadlock due to interoperating FSMs (cache, coherence unit, bus). The coherence stress test continues to work fine, and riscv-tests continue to pass.

One thing I want to change before merging is the hit logic for writes. Currently, there's an extremely nasty edge case where if a write request is started during a snoop hit, it will lower `proc_gen_bus_if.busy` before actually writing to the SRAM. This should only affect things which are able to lower `ren`/`wen` on the negative edge of the clock, however, I'm sure there's a better way to represent this logic that I can't think of right now.

### Performance

In merge sort, there is a ~5% speedup over the previous design. For the `global_atomic` toy test which tests atomic addition on a highly contentious variable, speedup is ~10%. There is likely still performance that can be squeezed out of the memory subsystem, however, its out of the scope of this PR.

### CODE REVIEW

File: verification/multicore-tests/c-tests/build_all.py [hunk 1]
```diff
@@ -20,34 +26,57 @@
...
+
+    print('Converting {} to binary'.format(filename))
+    rv = subprocess.run(cvt_cmd + [basename + '.elf', basename + '.bin'])
+    if rv.returncode != 0:
+        print('Exited with error {}, printing command, stdout, stderr!'.format(rv.returncode))
+        print('Command: {}\n\n'.format(compile_cmd + [filename, '-o', basename + '.elf']))
+        print('stdout:\n\n{}'.format(rv.stdout))
+        print('stderr:\n\n{}'.format(rv.stderr))
+        print('Exiting...')
+        exit(1)
```
Comment (devins2518) [inline verification/multicore-tests/c-tests/build_all.py:79]: Can you pull all this duplicated logic out to a function

File: source_code/caches/l1/l1_cache.sv [hunk 2]
```diff
@@ -320,7 +322,20 @@ module l1_cache #(
...
+                    casez (proc_gen_bus_if.byte_en)
+                        4'b0001:    proc_gen_bus_if.busy = sramRead.frames[hit_idx].data[decoded_addr.idx.block_bits][0+:8] != proc_gen_bus_if.wdata[0+:8];
+                        4'b0010:    proc_gen_bus_if.busy = sramRead.frames[hit_idx].data[decoded_addr.idx.block_bits][8+:8] != proc_gen_bus_if.wdata[8+:8];
+                        4'b0100:    proc_gen_bus_if.busy = sramRead.frames[hit_idx].data[decoded_addr.idx.block_bits][16+:8] != proc_gen_bus_if.wdata[16+:8];
+                        4'b1000:    proc_gen_bus_if.busy = sramRead.frames[hit_idx].data[decoded_addr.idx.block_bits][24+:8] != proc_gen_bus_if.wdata[24+:8];
+                        4'b0011:    proc_gen_bus_if.busy = sramRead.frames[hit_idx].data[decoded_addr.idx.block_bits][0+:16] != proc_gen_bus_if.wdata[0+:16];
+                        4'b1100:    proc_gen_bus_if.busy = sramRead.frames[hit_idx].data[decoded_addr.idx.block_bits][16+:16] != proc_gen_bus_if.wdata[16+:16];
+                        default:    proc_gen_bus_if.busy = sramRead.frames[hit_idx].data[decoded_addr.idx.block_bits] != proc_gen_bus_if.wdata;
+                    endcase
```
Comment (devins2518) [inline source_code/caches/l1/l1_cache.sv:338]: This still needs to be fixed

**Review by cole-nelson — APPROVED** [review]
LGTM.
